### ISSUE TITLE
Device-to-device slice write copies the source's base extent, not the view's size

### ISSUE THREAD

**Issue by tdoublep** [torch-spyre/torch-spyre#3826]
### What happened?

`dst[a:b] = src` writes the wrong number of rows when `src` is a prefix view of a longer tensor. It copies the view's base extent instead of the view's size, so it overwrites rows past the end of the destination slice.

Eager only. The compiled lowering of the same write is correct.

It also only happens after an earlier prefix-view slice write at a *different* view length has run in the same process. The first such write in a process is always correct.

### What did you expect to happen?

`dst[32:64] = src` should write 32 rows and leave the rest alone, as it does on CPU.

### How can we reproduce it?

fp16, AIU 1.0:

```python
import torch
import torch_spyre  # noqa: F401


def prefix_view_write(mode, qlen):
    """dst[32:32+qlen] = <first qlen rows of a 64-row tensor>.

    Returns (rows changed outside the window, window is correct).
    """

    def fn(dst, src):
        dst[32 : 32 + qlen] = src

    if mode == "compile":
        fn = torch.compile(fn, dynamic=False, backend="inductor")

    src_cpu = torch.randn(64, 32, 128, dtype=torch.float16)
    dst = torch.zeros(96, 32, 128, dtype=torch.float16).to("spyre")
    fn(dst, src_cpu.to("spyre")[:qlen])

    got = dst.cpu()
    outside = torch.ones(96, dtype=torch.bool)
    outside[32 : 32 + qlen] = False
    spill = int((got[outside].abs().amax(dim=(1, 2)) > 1e-3).sum())
    window_ok = torch.allclose(got[32 : 32 + qlen], src_cpu[:qlen], atol=1e-2)
    return spill, window_ok


for mode in ("eager", "compile"):
    print(mode, "warm-up:", prefix_view_write(mode, 16), "under test:", prefix_view_write(mode, 32))
```

Output:

```
eager warm-up: (0, True) under test: (32, True)
compile warm-up: (0, True) under test: (0, True)
```

In eager the second call asks for 32 rows but writes 64, so `dst[64:96]` receives `src`'s rows 32..63. The window `dst[32:64]` stays correct in every case, so only the extra rows are wrong.

Both calls are needed. The second one on its own is correct, in either mode.

### Anything else we need to know?

| source | eager | compile |
|---|---|---|
| exact-size tensor | ok | ok |
| prefix view, first write in the process | ok | ok |
| prefix view, after a prefix-view write at a different length | writes the base extent | ok |
| prefix view + `.contiguous()` | writes the base extent | ok |
| prefix view + `.clone()` | ok | ok |

- The write always ends at `offset + base_extent`.
- `.contiguous()` changes nothing: a prefix slice is already contiguous and keeps the same storage. `.clone()` shrinks the extent and fixes it.
- What triggers it is a change in the source view length. Changing the destination size or the write offset does not.
- Where the overrun runs past the end of the destination tensor, the write leaves the tensor entirely.
- CPU is fine. Same behaviour for `dst[a:b] = src` and `narrow().copy_()`.
- No `FallbackWarning` in either mode, so the compiled case is genuinely correct on device rather than quietly running on CPU.
- The eager path goes through `spyre__copy_from` -> `torch.ops.spyre.copy_from_d2d`, which forwards storage offsets but nothing about the views' sizes.

Found in spyre-inference paged attention: the write-back sliced padding rows off the result, and the over-long write corrupted other sequences in the batch and produced NaN. The write-back is eager, so it is exposed. Worked around with `.clone()` in torch-spyre/spyre-inference#546.

**Comment by tdoublep** [comment]
Confirmed reproducing on an AIU 1.0 pod (2 cards, DD2), fp16, with the reproducer from the issue body run verbatim. Output matches the report exactly, and is identical across two consecutive runs in fresh processes:

```
eager   warm-up: (0, True)  under test: (32, True)
compile warm-up: (0, True)  under test: (0, True)
```

### Extra data point: the spill size tracks the *base* extent, not the view

Sweeping the warm-up / under-test view lengths within one process (so every pair after the first is already "after a prefix-view write at a different length"):

| warm-up qlen | test qlen | spilled rows | window ok |
|---|---|---|---|
| 16 | 32 | 32 | yes |
| 32 | 16 | **48** | yes |
| 8  | 64 | 0  | yes |
| 64 | 32 | 32 | yes |
| 16 | 64 | 0  | yes |
| 48 | 32 | 32 | yes |

Every row is consistent with "the write always ends at `offset + base_extent`" = `32 + 64` = 96, independent of the requested view length:

- `qlen=16` → window is `dst[32:48]`, write reaches row 96 → 48 rows spilled.
- `qlen=32` → window is `dst[32:64]`, write reaches row 96 → 32 rows spilled.
- `qlen=64` → window is `dst[32:96]`, which already ends at 96 → 0 spilled, so this length can never expose the bug on a 96-row destination.

The window itself is correct in every case, as reported.

### Environment

Host / stack:

```
hostname:        tpa-spyre-dev-2
kernel:          Linux-5.14.0-570.131.1.el9_6.x86_64
os:              Red Hat Enterprise Linux 10.2 (Coughlan)
python:          3.12.13
torch:           2.13.0+cpu
torch-spyre:     0.0.1 @ d83995da4dfabcee036bd17ad60a73124a789a0b
vllm:            0.27.1+empty @ 6e448d0ea9bf3d88d898b65449ca6dc2aec170ac
spyre-inference: 0.1.dev276 (main @ f44a004)
numpy:           2.3.5
```

The installed `torch-spyre` is clean — no hand-applied local patches. All 283 installed files were hashed and compared against the wheel's `RECORD`, and every one matches the recorded install exactly (`torch` likewise, 12251 files). That rev is also the rev pinned in `spyre-inference`'s `pyproject.toml`, so this is a stock build of `d83995da`.

Installed `ibm-*` RPMs:

```
ibm-deeptools-2.0.0-0.main.1+2241.1be8bc9_279.el10
ibm-deeptools-devel-2.0.0-0.main.1+2241.1be8bc9_279.el10
ibm-senlib-core-2.0.0-0.main.1+235.7f44c00_198.el10
ibm-senlib-dd2-2.0.0-0.main.1+235.7f44c00_198.el10
ibm-senlib-headers-2.0.0-0.main.1+235.7f44c00_198.el10
ibm-flex-2.0.0-0.main.1+473.f97235a_283.el10
ibm-flex-devel-2.0.0-0.main.1+473.f97235a_283.el10
ibm-libaiupti-2.0.0-0.main.1+18.98f7d31_0.el10
ibm-spyre-comms-1.0.0-0.main.1+119.fc4c8d1_136.el10
ibm-spyre-comms-devel-1.0.0-0.main.1+119.fc4c8d1_136.el10
ibm-aiu-toolbox-e2e-2.0.0-0.main.1+27.4f08606_80.el10
```

`/opt/ibm/spyre/components.txt`:

```
ibm-deeptools:2.0.0-0.main.1+2241.1be8bc9_279.el10
ibm-senlib-core:2.0.0-0.main.1+235.7f44c00_198.el10
ibm-senlib-dd2:2.0.0-0.main.1+235.7f44c00_198.el10
ibm-flex:2.0.0-0.main.1+473.f97235a_283.el10
ibm-aiu-toolbox-e2e:2.0.0-0.main.1+27.4f08606_80.el10
```

Cards:

```
0000:aa:00.0  device_id=06a7  boot=9  RPD 1100.0 MHz / SOC 1000.0 MHz  112.0 GB  serial=FZ84V-03-07-11
0000:ab:00.0  device_id=06a7  boot=9  RPD 1100.0 MHz / SOC 1000.0 MHz  112.0 GB  serial=FZ859-14-07-12
```

Relevant env: `FLEX_COMPUTE=SENTIENT`, `FLEX_DEVICE=PF`, `AIU_WORLD_SIZE=2`, `TORCH_SENDNN_CACHE_ENABLE=1`, `DT_DEEPRT_VERBOSE=-1`. No `FallbackWarning` in either mode, so both paths are genuinely running on device.
